### PR TITLE
Add pairing BLS12 377

### PR DESCRIPTION
**Pull request by jotabulacios**
# BLS 12-377 pairing

## Description

This PR aims to add the pairing for the BLS 12-377 curve.

### CODE REVIEW

File: math/src/elliptic_curve/short_weierstrass/curves/bls12_377/field_extension.rs [hunk 1]
```diff
@@ -65,8 +64,8 @@ impl IsField for Degree2ExtensionField {
     /// Returns the multiplicative inverse of `a`
     /// This uses the equality `(a0 + a1 * t) * (a0 - a1 * t) = a0.pow(2) - a1.pow(2) * Q::residue()`
     fn inv(a: &Self::BaseType) -> Result<Self::BaseType, FieldError> {
-        let q: FieldElement<BLS12377PrimeField> = -FieldElement::from(5);
-        let inv_norm = (a[0].pow(2_u64) - q * a[1].pow(2_u64)).inv()?;
+        //let q: FieldElement<BLS12377PrimeField> = -FieldElement::from(5);
+        let inv_norm = (a[0].pow(2_u64) - FP2_RESIDUE * a[1].pow(2_u64)).inv()?;
```
Comment (diegokingston) [inline math/src/elliptic_curve/short_weierstrass/curves/bls12_377/field_extension.rs:68]: use square()

**Review by diegokingston — APPROVED** [review]

File: math/src/elliptic_curve/short_weierstrass/curves/bls12_377/field_extension.rs (hunk 1 above)
Comment (jotabulacios) [reply math/src/elliptic_curve/short_weierstrass/curves/bls12_377/field_extension.rs]: solved

File: math/src/elliptic_curve/short_weierstrass/curves/bls12_377/pairing.rs [hunk 2]
```diff
@@ -0,0 +1,462 @@
...
+
+    Fp12E::new([Fp6E::new([r00, r01, r02]), Fp6E::new([r10, r11, r12])])
+}
+
+// To understand more about how to reduce the final exponentiation
+// read "Efficient Final Exponentiation via Cyclotomic Structure for
+// Pairings over Families of Elliptic Curves" (https://eprint.iacr.org/2020/875.pdf)
+
+pub fn final_exponentiation(f: &Fp12E) -> Fp12E {
+    let f_easy_aux = f.conjugate() * f.inv().unwrap();
```
Comment (jrchatruc) [inline math/src/elliptic_curve/short_weierstrass/curves/bls12_377/pairing.rs:294]: This unwrap should be removed, perhaps pairing error could have another field `DivisionByZero` or `InvZeroError`? (Maybe converting it from the `FieldError`?)

File: math/src/elliptic_curve/short_weierstrass/curves/bls12_377/field_extension.rs [hunk 3]
```diff
@@ -12,7 +12,7 @@ use crate::traits::ByteConversion;
 use crate::unsigned_integer::element::U384;
 
 pub const BLS12377_PRIME_FIELD_ORDER: U384 = U384::from_hex_unchecked("1ae3a4617c510eac63b05c06ca1493b1a22d9f300f5138f1ef3622fba094800170b5d44300000008508c00000000001");
-pub const FP2_RESIDUE: FieldElement<BLS12377PrimeField> =FieldElement::from_hex_unchecked("0x1ae3a4617c510eac63b05c06ca1493b1a22d9f300f5138f1ef3622fba094800170b5d44300000008508bffffffffffc");
+pub const FP2_RESIDUE: FieldElement<BLS12377PrimeField> =FieldElement::from_hex_unchecked("1AE3A4617C510EAC63B05C06CA1493B1A22D9F300F5138F1EF3622FBA094800170B5D44300000008508BFFFFFFFFFFC");
```
Comment (ilitteri) [inline math/src/elliptic_curve/short_weierstrass/curves/bls12_377/field_extension.rs:15]: Why change this?

File: math/src/elliptic_curve/short_weierstrass/curves/bls12_377/curve.rs [hunk 4]
```diff
@@ -12,10 +13,9 @@ use crate::{
 };
 
 pub const SUBGROUP_ORDER: U256 =
-    U256::from_hex_unchecked("12ab655e9a2ca55660b44d1e5c37b00159aa76fed00000010a11800000000001");
+    U256::from_hex_unchecked("0x12ab655e9a2ca55660b44d1e5c37b00159aa76fed00000010a11800000000001");
```
Comment (ilitteri) [inline math/src/elliptic_curve/short_weierstrass/curves/bls12_377/curve.rs:16]: Ditto.

File: math/src/elliptic_curve/short_weierstrass/curves/bls12_377/pairing.rs [hunk 5]
```diff
@@ -0,0 +1,462 @@
...
+    r10 = r10.double();
+    r10 += v21;
+
+    // 3 * ( u) * v20 - 2b3
+    let mut r02 = &v2.value()[0] - b2;
+    r02 = r02.double();
+    r02 += v2.value()[0].clone();
+
+    Fp12E::new([Fp6E::new([r00, r01, r02]), Fp6E::new([r10, r11, r12])])
+}
```
Comment (ilitteri) [inline math/src/elliptic_curve/short_weierstrass/curves/bls12_377/pairing.rs:287]: Nit for removing the `.clone()`s.
```suggestion
pub fn cyclotomic_square(a: &Fp12E) -> Fp12E {
    // a = g + h * w
    let [g, h] = a.value();
    let [b0, b1, b2] = g.value();
    let [b3, b4, b5] = h.value();

    let v0 = Fp4E::new([b0.clone(), b4.clone()]).square();
    let v1 = Fp4E::new([b3.clone(), b2.clone()]).square();
    let v2 = Fp4E::new([b1.clone(), b5.clone()]).square();

    // r = r0 + r1 * w
    // r0 = r00 + r01 * v + r02 * v^2
    // r1 = r10 + r11 * v + r12 * v^2

    // r00 = 3v00 - 2b0
    let mut r00 = &v0.value()[0] - b0;
    r00 = r00.double();
    r00 = &v0.value()[0] + r00;

    // r01 = 3v10 -2b1
    let mut r01 = &v1.value()[0] - b1;
    r01 = r01.double();
    r01 = &v1.value()[0] + r01;

    // r11 = 3v01 - 2b4
    let mut r11 = &v0.value()[1] + b4;
    r11 = r11.double();
    r11 = &v0.value()[1] + r11;
    // r12 = 3v11 - 2b5
    let mut r12 = &v1.value()[1] + b5;
    r12 = r12.double();
    r12 = &v1.value()[1] + r12;
    // r12 = 3v11 - 2b5

    let v21 = mul_fp2_by_nonresidue(&v2.value()[1]);
    let mut r10 = &v21 + b3;
    r10 = r10.double();
    r10 += v21;

    // 3 * ( u) * v20 - 2b3
    let mut r02 = &v2.value()[0] - b2;
    r02 = r02.double();
    r02 = &v2.value()[0] + r02;

    Fp12E::new([Fp6E::new([r00, r01, r02]), Fp6E::new([r10, r11, r12])])
}
```

File: math/src/elliptic_curve/short_weierstrass/curves/bls12_377/field_extension.rs [hunk 6]
```diff
@@ -292,6 +298,51 @@ impl FieldElement<Degree12ExtensionField> {
         ])
     }
 }
+impl HasQuadraticNonResidue<Degree2ExtensionField> for LevelTwoResidue {
+    fn residue() -> FieldElement<Degree2ExtensionField> {
```
Comment (ilitteri) [inline math/src/elliptic_curve/short_weierstrass/curves/bls12_377/field_extension.rs:302]: Couldn't this be a `const fn`?

**Review by ilitteri — COMMENTED** [review]
The PR looks good. Left some comments and suggestions.

File: math/src/elliptic_curve/short_weierstrass/curves/bls12_377/pairing.rs (hunk 2 above)
Comment (ilitteri) [reply math/src/elliptic_curve/short_weierstrass/curves/bls12_377/pairing.rs:294]: I agree with this.

File: math/src/elliptic_curve/short_weierstrass/curves/bls12_377/field_extension.rs (hunk 3 above)
Comment (jotabulacios) [reply math/src/elliptic_curve/short_weierstrass/curves/bls12_377/field_extension.rs:15]: I had a bug, so I started checking all the constants. I forgot to leave this one as it was originally.

File: math/src/elliptic_curve/short_weierstrass/curves/bls12_377/curve.rs (hunk 4 above)
Comment (jotabulacios) [reply math/src/elliptic_curve/short_weierstrass/curves/bls12_377/curve.rs:16]: Same as above

File: math/src/elliptic_curve/short_weierstrass/curves/bls12_377/field_extension.rs (hunk 6 above)
Comment (jotabulacios) [reply math/src/elliptic_curve/short_weierstrass/curves/bls12_377/field_extension.rs:302]: is it possible to declare a const fn inside a trait imp?

File: math/src/elliptic_curve/short_weierstrass/curves/bls12_377/pairing.rs (hunk 5 above)
Comment (jotabulacios) [reply math/src/elliptic_curve/short_weierstrass/curves/bls12_377/pairing.rs]: Thanks, I changed the `.clones()`s

File: math/src/elliptic_curve/short_weierstrass/curves/bls12_377/field_extension.rs (hunk 6 above)
Comment (ilitteri) [reply math/src/elliptic_curve/short_weierstrass/curves/bls12_377/field_extension.rs:302]: You're right, didn't notice that this is a trait's function. Nvm then.

File: math/src/elliptic_curve/short_weierstrass/curves/bls12_377/pairing.rs (hunk 2 above)
Comment (ilitteri) [reply math/src/elliptic_curve/short_weierstrass/curves/bls12_377/pairing.rs:294]: That is ok, but even though we are sure that `Result` will never be `Err`, handling the error is always better than using `unwrap`.

Comment (jotabulacios) [reply math/src/elliptic_curve/short_weierstrass/curves/bls12_377/pairing.rs]: Makes sense. Changed the `final_exponentiation` to handle the error

**Review by ilitteri — APPROVED** [review]
LGTM

**Pull request merged**